nft_ingester/src/program_transformers/asset_upserts.rs

```diff
@@ -0,0 +1,153 @@
+    query.sql);
+    txn_or_conn.execute(query).await?;
+    Ok(())
+}
+
+pub struct AssetMetadataAccountColumns {
+    pub mint: Vec<u8>,
+    pub metadata_account_id: Vec<u8>,
+    pub owner_type: OwnerType,
+    pub specification_asset_class: Option<SpecificationAssetClass>,
+    pub royalty_amount: i32,
+    pub asset_data: Option<Vec<u8>>,
+    pub slot_updated_metadata_account: u64,
+}
+
+pub async fn upsert_assets_metadata_account_columns<T: ConnectionTrait + TransactionTrait>(
+    columns: AssetMetadataAccountColumns,
+    txn_or_conn: &T,
+) -> Result<(), DbErr> {
+    let active_model = asset::ActiveModel {
+        id: Set(columns.mint),
+        owner_type: Set(columns.owner_type),
+        specification_version: Set(Some(
+            digital_asset_types::dao::sea_orm_active_enums::SpecificationVersions::V1,
+        )),
+        specification_asset_class: Set(columns.specification_asset_class),
+        tree_id: Set(None),
+        nonce: Set(Some(0)),
+        seq: Set(Some(0)),
+        leaf: Set(None),
+        data_hash: Set(None),
+        creator_hash: Set(None),
+        compressed: Set(false),
+        compressible: Set(false),
+        royalty_target_type: Set(RoyaltyTargetType::Creators),
+        royalty_target: Set(None),
+        royalty_amount: Set(columns.royalty_amount),
+        asset_data: Set(columns.asset_data),
+        slot_updated_metadata_account: Set(Some(columns.slot_updated_metadata_account as i64)),
+        burnt: Set(false),
+        ..Default::default()
+    };
+    let mut query = asset::Entity::insert(active_model)
+        .on_conflict(
+            OnConflict::columns([asset::Column::Id])
+                .update_columns([
+                    asset::Column::OwnerType,
+                    asset::Column::SpecificationVersion,
+                    asset::Column::SpecificationAssetClass,
+                    asset::Column::TreeId,
+                    asset::Column::Nonce,
+                    asset::Column::Seq,
+                    asset::Column::Leaf,
+                    asset::Column::DataHash,
+                    asset::Column::CreatorHash,
+                    asset::Column::Compressed,
+                    asset::Column::Compressible,
+                    asset::Column::RoyaltyTargetType,
+                    asset::Column::RoyaltyTarget,
+                    asset::Column::RoyaltyAmount,
+                    asset::Column::AssetData,
+                    asset::Column::SlotUpdatedMetadataAccount,
+                    asset::Column::Burnt,
+                ])
+                .to_owned(),
+        )
+        .build(DbBackend::Postgres);
+
+    query.sql = format!(
+        "{} WHERE excluded.slot_updated_metadata_account >= asset.slot_updated_metadata_account OR asset.slot_updated_metadata_account IS NULL",
+        query.sql);
+    txn_or_conn.execute(query).await?;
```

Comment:
I do see the transaction being used to batch all these updates into a single statement. This may suffice for reducing interactions with the db but still may cause delays as transactions require more aggressive locking but if others aren't accessing the same rows then should be ok.

Reply:
It looks like in `save_v1_asset` that `index_and_fetch_mint_data` and `index_token_account_data` take the `conn` object, so in `save_v1_asset` the 3 calls are not batched into one transaction.  That makes sense to me because those the `index_` functions have queries with retries and probably because of that don't belong in a single transaction.

I will say separately that we followed a similar strategy on compressed NFTs with using multiple `seq` numbers to protect the different scoped data for compression, and putting multiple upserts into single transactions had measurable effect on the indexing.  I measured it and saw significant reduction on the order of like 9-23 ms down to 204 ms for 5 upserts to the asset table in Bubblegum `mint_v1` (reference: https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/pull/142).  Later @NicolasPennie put the remaining upserts from Bubblegum `mint_v1` (`asset_data`, `asset_creators`, etc.) into a single transaction and it performed better in his metrics.